### PR TITLE
Standardize record puns (#867)

### LINKED ISSUE #867: Record puns
I'd like to propose new syntax for record literals:

```
{ key } = { key = key }
```

So, for example:

```
let key = "value" in { key } : { key : Text }
```

I think this greatly reduces noise when returning a big record.

### PR DESCRIPTION
**Pull request by Nadrieril**
Fixes #867

Matching change to the Rust implementation: https://github.com/Nadrieril/dhall-rust/pull/143

I also snuck in a cosmetic change to the `with` grammar for those of us with auto-generated parsers.

### CODE REVIEW

File: tests/type-inference/success/unit/RecordLitPunSomeA.dhall [hunk 1]
```diff
@@ -0,0 +1 @@
+let `None` = 0 let `Some` = \(_: Bool) -> 1 in { None, Some }
```
Comment (sjakobi) [inline tests/type-inference/success/unit/RecordLitPunSomeA.dhall]: Nice!

I hadn't expected puns to be usable for `Some` fields!

**Comment by Profpatsch** [comment]
So `{ x.y.z }` desugars to `{ x = { y = { z = z } } }`?

**Comment by Gabriel439** [comment]
@Profpatsch: It would be a parse error, according to this test:

https://github.com/dhall-lang/dhall-lang/pull/938/files#diff-dc8f2eb8726e4dbc33b8472447ed3266

**Comment by jvanbruegge** [comment]
I would probably expect it to either be `{ z = x.y.z }` or `{ x = { y = { z = x.y.z } } }`

**Pull request merged**